File: solutions/javascript/mixed-juices/1/mixed-juices.js
// @ts-check

/**
 * Determines how long it takes to prepare a certain juice.
 *
 * @param {string} name
 * @returns {number} time in minutes
 */
export function timeToMixJuice(name) {
    console.log(name);
    switch (name) {
        case 'All or Nothing':
            return 5;
        case 'Energizer':
        case 'Green Garden':
            return 1.5;
        case 'Pure Strawberry Joy':
            return 0.5;
        case 'Tropical Island':
            return 3;
        default:
            return 2.5;
    }
}

/**
 * Calculates the number of limes that need to be cut
 * to reach a certain supply.
 *
 * @param {number} wedgesNeeded
 * @param {string[]} limes
 * @returns {number} number of limes cut
 */
export function limesToCut(wedgesNeeded, limes) {
    let limesCut = 0;

    while (wedgesNeeded > 0) {
        let currentLime = limes.shift();
        if (!currentLime) break;

        switch (currentLime) {
            case 'small':
                wedgesNeeded -= 6;
                break;
            case 'medium':
                wedgesNeeded -= 8;
                break;
            case 'large':
                wedgesNeeded -= 10;
                break;
        }
        console.log(wedgesNeeded);
        limesCut++;
    }

    return limesCut;
}

/**
 * Determines which juices still need to be prepared after the end of the shift.
 *
 * @param {number} timeLeft
 * @param {string[]} orders
 * @returns {string[]} remaining orders after the time is up
 */
export function remainingOrders(timeLeft, orders) {
    while (timeLeft > 0) {
        let currentOrder = orders.shift();
        if (!currentOrder) break;
        timeLeft -= timeToMixJuice(currentOrder);
    }

    return orders;
}

File: solutions/julia/mixed-juices/1/mixed-juices.jl
function time_to_mix_juice(juice)
    juice == "Pure Strawberry Joy" && return 0.5
    juice == "Energizer" && return 1.5
    juice == "Green Garden" && return 1.5
    juice == "Tropical Island" && return 3.0
    juice == "All or Nothing" && return 5.0
    return 2.5
end

function wedges_from_lime(size)
    size == "small" && return 6
    size == "medium" && return 8
    size == "large" && return 10
end

function limes_to_cut(needed, limes)
    wedges_cut = 0
    limes_cut = 0

    while wedges_cut < needed
        length(limes) == 0 && break
        wedges_cut += wedges_from_lime(popfirst!(limes))
        limes_cut += 1
    end

    return limes_cut
end

function order_times(orders)
    times = []

    for order in orders
        push!(times, time_to_mix_juice(order))
    end

    return times
end

function remaining_orders(time_left, orders)
    time_spent = 0

    while time_spent < time_left
        length(orders) == 0 && break
        time_spent += time_to_mix_juice(popfirst!(orders))
    end

    return orders
end
